### PR TITLE
Improve IntelliSense support when using VS Code

### PR DESCRIPTION
**Pull request by noisysocks**
Improves IntelliSense support in Visual Studio Code by providing a [`jsconfig.json` file](https://code.visualstudio.com/docs/languages/jsconfig) that correctly maps `@wordpress/` paths to their source location.

This lets developers cmd+click on symbols to navigate to where they're defined:

navigation

It also improves the hints shown when using symbols imported from `@wordpress/` paths:

hints

### CODE REVIEW

File: jsconfig.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+			"@wordpress/*": ["./*", "./packages/*/src"]
+		}
+	},
+	"exclude": [
+		"build",
+		"build-module",
+		"coverage",
+		"node_modules",
+		"test/e2e/test-plugins",
+		"vendor"
```
Comment (noisysocks) [inline jsconfig.json]: I borrowed these excludes from `.eslintignore`—not sure how up to date they are. From what I understand reading [the docs](https://code.visualstudio.com/docs/languages/jsconfig), VS Code just recommends that you exclude directories with lots of irrelevant source files e.g. `build` and `node_modules`.

**Comment by gziolo** [comment]
Nice stuff. I'm wondering if it should be a local file which you add to the list of ignored files using your local bash configuration.

**Comment by ntwb** [comment]
Agreed, it should be a local file added to your local Git `.gitignore` file

**Comment by noisysocks** [comment]
For what it's worth, [Calypso has a `jsconfig.json`](https://github.com/Automattic/wp-calypso/blob/master/jsconfig.json) which is very useful.

> Agreed, it should be a local file added to your local Git .gitignore file

I don't think adding `jsconfig.json` to `.gitignore` makes sense because it's not automatically generated like `.idea` is. It contains information about the JavaScript project, not editor configuration settings.

> unless it isn't specific only to VS Code

It's not specific to VS Code in that any tool is free to parse it 😉 For example, [Sourcegraph uses `jsconfig.json`](https://about.sourcegraph.com/docs/code-intelligence/javascript/)  to power JavaScript hinting in their [Atom plugin](https://atom.io/packages/sourcegraph) and [Sublime Text plugin](https://github.com/sourcegraph/sourcegraph-sublime).

**Comment by noisysocks** [comment]
Ask yourself: which option does more good for the world?

1) We add `jsconfig.json`. VS Code users get a useful productivity gain for free. Other developers go on about their days and don't notice a thing.
2) We don't add `jsconfig.json`. A few VS Code users spend the necessary 1-2 hours as I did learning about `jsconfig.json` and setting it up. Most VS Code developers go on about their days and don't know what they're missing out on. Other developers go on about their days and don't notice a thing.

**Comment by pento** [comment]
I'm in two minds about this.

On the one hand, I'm not wild about us adding IDE-specific stuff to Core. It could pretty easily get out of hand, if everyone wants the config for their specific editor added.

On the other hand, we have the `.editorconfig` file, but the Editorconfig spec doesn't cover a bunch of useful stuff, and it barely moves, whereas IDE maintainers are far more likely to keep their stuff moving.

It's unfortunately that a good, global editor config standard hasn't really emerged, that would make the decision a bunch easier.

I'm inclined to put this in Gutenberg for now, and look at how we can practically allow IDE configs in Core. Perhaps base it on [current popularity](https://insights.stackoverflow.com/survey/2017#technology-most-popular-developer-environments-by-occupation) or something.

**Comment by ntwb** [comment]
_I didn't expect `Notepad++` to rank so highly there..._

Thanks @noisysocks, let's do it under the guise that `jsconfig.json` is *not* Editor/IDE specific 😉

**Review by ntwb — APPROVED** [review]

**Comment by noisysocks** [comment]
> It's unfortunately that a good, global editor config standard hasn't really emerged, that would make the decision a bunch easier.

_“Imagine a world with one global editor config standard.”_ 

iur

**Pull request merged**